File: src/main/java/com/ternbusty/takoyaki/nativeimage/GeneratedForeignFeature.java
package com.ternbusty.takoyaki.nativeimage;

import org.graalvm.nativeimage.hosted.Feature;
import org.graalvm.nativeimage.hosted.RuntimeForeignAccess;

import java.lang.foreign.FunctionDescriptor;
import java.lang.reflect.Field;
import java.lang.reflect.Modifier;

/**
 * Register every downcall the jextract-generated bindings declare.
 *
 * jextract emits one nested class per function holding a static
 * {@code FunctionDescriptor DESC}, so the whole set can be harvested by
 * reflection instead of being restated by hand in {@link ForeignFeature}.
 * Adding a function to a jextract task therefore needs no registration change.
 */
public final class GeneratedForeignFeature implements Feature {

    private static final String BASE_CLASS = "com.ternbusty.takoyaki.syscall.gen.NativeHRaw";

    @Override
    public void duringSetup(DuringSetupAccess access) {
        int registered = 0;
        ClassLoader cl = getClass().getClassLoader();

        for (int i = 0; ; i++) {
            String name = (i == 0) ? BASE_CLASS : BASE_CLASS + "_" + i;
            Class<?> header;
            try {
                header = Class.forName(name, false, cl);
            } catch (ClassNotFoundException e) {
                if (i == 0) throw new RuntimeException("base header class not found: " + BASE_CLASS, e);
                break;
            }
            registered += scanHeader(header);
        }
        System.out.println("[GeneratedForeignFeature] registered " + registered + " downcalls");
    }

    private static int scanHeader(Class<?> header) {
        int count = 0;
        for (Class<?> fn : header.getDeclaredClasses()) {
            for (Field f : fn.getDeclaredFields()) {
                if (f.getType() != FunctionDescriptor.class
                        || !Modifier.isStatic(f.getModifiers())) {
                    continue;
                }
                try {
                    f.setAccessible(true);
                    RuntimeForeignAccess.registerForDowncall((FunctionDescriptor) f.get(null));
                    count++;
                } catch (Throwable t) {
                    // A descriptor we cannot read is left to ForeignFeature.
                }
            }
        }
        return count;
    }
}

File: src/main/java/com/ternbusty/takoyaki/nativeimage/GeneratedForeignFeature.kt
package com.ternbusty.takoyaki.nativeimage

import org.graalvm.nativeimage.hosted.Feature
import org.graalvm.nativeimage.hosted.Feature.DuringSetupAccess
import org.graalvm.nativeimage.hosted.RuntimeForeignAccess

import java.lang.foreign.FunctionDescriptor
import java.lang.reflect.Modifier

/**
 * Register every downcall the jextract-generated bindings declare.
 *
 * jextract emits one nested class per function holding a static
 * `FunctionDescriptor DESC`, so the whole set can be harvested by
 * reflection instead of being restated by hand in [ForeignFeature].
 * Adding a function to a jextract task therefore needs no registration change.
 */
class GeneratedForeignFeature : Feature {

    override fun duringSetup(access: DuringSetupAccess) {
        var registered = 0
        // jextract may split a large header into NativeH, NativeH_1, NativeH_2, ...
        // Iterate until we hit a class that does not exist.
        var i = 0
        while (true) {
            val name = if (i == 0) BASE_CLASS else "${BASE_CLASS}_$i"
            val header = try {
                Class.forName(name, false, javaClass.classLoader)
            } catch (_: ClassNotFoundException) {
                break
            }
            registered += scanHeader(header)
            i++
        }
        println("[GeneratedForeignFeature] registered $registered downcalls")
    }

    private fun scanHeader(header: Class<*>): Int {
        var count = 0
        for (fn in header.declaredClasses) {
            for (f in fn.declaredFields) {
                if (f.type != FunctionDescriptor::class.java
                    || !Modifier.isStatic(f.modifiers)
                ) {
                    continue
                }
                try {
                    f.isAccessible = true
                    RuntimeForeignAccess.registerForDowncall(f.get(null) as FunctionDescriptor)
                    count++
                } catch (_: Throwable) {
                    // A descriptor we cannot read is left to ForeignFeature.
                }
            }
        }
        return count
    }

    companion object {
        /** Base class name for jextract-generated headers. */
        private const val BASE_CLASS = "com.ternbusty.takoyaki.syscall.gen.NativeH"
    }
}
